--!strict
--[=[
	Optimized these functions for speed as well as preserving fidality.
	In the future, use Roblox's orthogonal angle format.

	@class CFrameSerializationUtils
]=]

const HttpService = game:GetService("HttpService")

const MathUtils = require("./MathUtils")

const CFrameSerializationUtils = {}

const PRECISION = 10000
const MULTIPLIER = 128

export type SerializedCFrame = { number }

--[=[
	Outputs the rotation
	@param cf CFrame
	@return { number, number, number, number, number, number }
]=]
@checked
function CFrameSerializationUtils.outputRotationAzure(cf: CFrame): SerializedCFrame
	const lookVector = cf.LookVector
	local azumith = math.atan2(-lookVector.X, -lookVector.Z)
	const ybase = (lookVector.X ^ 2 + lookVector.Z ^ 2) ^ 0.5
	local elevation = math.atan2(lookVector.Y, ybase)

	const withoutRoll = CFrame.Angles(0, azumith, 0) * CFrame.Angles(elevation, 0, 0) + cf.Position
	local _, _, roll = (withoutRoll:Inverse() * cf):ToEulerAnglesXYZ()

	-- Atan2 -> in the range [-pi, pi]
	azumith = MathUtils.round((azumith / math.pi) * PRECISION)
	roll = MathUtils.round((roll / math.pi) * PRECISION)
	elevation = MathUtils.round((elevation / (math.pi / 2)) * PRECISION)

	-- Buffer:WriteSigned(22, azumith)
	-- Buffer:WriteSigned(21, roll)
	-- Buffer:WriteSigned(21, elevation)

	local px, py, pz = cf.X, cf.Y, cf.Z
	px = MathUtils.round(px * MULTIPLIER)
	py = MathUtils.round(py * MULTIPLIER)
	pz = MathUtils.round(pz * MULTIPLIER)
	return { px, py, pz, azumith, roll, elevation }
end

--[=[
	Encodes a CFrame into JSON for serialization in attributes.

	@param cf CFrame
	@return string
]=]
@checked
function CFrameSerializationUtils.toJSONString(cf: CFrame): string
	return HttpService:JSONEncode(CFrameSerializationUtils.outputRotationAzure(cf))
end

--[=[
	Returnst true if it's a table encoded cframe

	@param data any
	@return boolean
]=]
@checked
function CFrameSerializationUtils.isRotationAzure(data: any): boolean
	return type(data) == "table"
		and type(data[1]) == "number"
		and type(data[2]) == "number"
		and type(data[3]) == "number"
		and type(data[4]) == "number"
		and type(data[5]) == "number"
		and type(data[6]) == "number"
end

--[=[
	Decodes a CFrame from JSON. For serialization in attributes.

	@param str string
	@return CFrame
]=]
@checked
function CFrameSerializationUtils.fromJSONString(str: string): CFrame?
	const decoded = HttpService:JSONDecode(str)
	if CFrameSerializationUtils.isRotationAzure(decoded) then
		return CFrameSerializationUtils.readRotationAzure(decoded)
	else
		return nil
	end
end

--[=[
	Returns the position
	@param data { number, number, number, number, number, number }
	@return Vector3
]=]
@checked
function CFrameSerializationUtils.readPosition(data: SerializedCFrame): Vector3
	return Vector3.new(data[1] / MULTIPLIER, data[2] / MULTIPLIER, data[3] / MULTIPLIER)
end

--[=[
	Returns the CFrame

	@param data { number, number, number, number, number, number }
	@return CFrame
]=]
@checked
function CFrameSerializationUtils.readRotationAzure(data: SerializedCFrame): CFrame
	local azumith = data[4]
	local roll = data[5] -- Buffer:ReadSigned(21)
	local elevation = data[6] -- Buffer:ReadSigned(21)

	azumith = math.pi * (azumith / PRECISION)
	roll = math.pi * (roll / PRECISION)
	elevation = (math.pi / 2) * (elevation / PRECISION)

	const cframe = CFrame.Angles(0, azumith, 0) * CFrame.Angles(elevation, 0, roll)
	return cframe + Vector3.new(data[1] / MULTIPLIER, data[2] / MULTIPLIER, data[3] / MULTIPLIER) --, azumith, roll, elevation}
end

return CFrameSerializationUtils
