--!strict
--[=[
	@class MessagingServiceUtils
]=]

const HttpService = game:GetService("HttpService")
const MessagingService = game:GetService("MessagingService")

const Promise = require("../roblox_packages/Promise")

const DEBUG_PUBLISH = false
const DEBUG_SUBSCRIBE = false

const MessagingServiceUtils = {}

--[=[
	Wraps MessagingService:PublishAsync(topic, message)
	@param topic string
	@param message any
	@return Promise
]=]
@checked
function MessagingServiceUtils.promisePublish(topic: string, message: any?)
	assert(type(topic) == "string", "Bad topic")

	return Promise.new(function(resolve, reject)
		if DEBUG_PUBLISH then
			print(string.format("Publishing on %q: ", topic), HttpService:JSONEncode(message))
		end

		const ok, err = pcall(function()
			MessagingService:PublishAsync(topic, message)
		end)
		if not ok then
			if DEBUG_PUBLISH then
				warn(string.format("Failed to publish on %q due to %q", topic, err or "nil"))
			end
			return reject(err)
		end
		return resolve()
	end)
end

--[=[
	Wraps MessagingService:SubscribeAsync(topic, callback)
	@param topic string
	@param callback callback
	@return Promise<RBXScriptConnection>
]=]
@checked
function MessagingServiceUtils.promiseSubscribe(topic: string, callback: (...any) -> ...any)
	assert(type(topic) == "string", "Bad topic")
	assert(type(callback) == "function", "Bad callback")

	if DEBUG_SUBSCRIBE then
		print(string.format("Listening on %q", topic))

		const oldCallback = callback
		callback = function(message)
			print(string.format("Recieved on %q", topic), HttpService:JSONEncode(message))
			oldCallback(message)
		end
	end

	return Promise.new(function(resolve, reject)
		local connection
		const ok, err = pcall(function()
			connection = MessagingService:SubscribeAsync(topic, callback)
		end)
		if not ok then
			if DEBUG_PUBLISH then
				warn(string.format("Failed to subscribe on %q due to %q", topic, err or "nil"))
			end
			return reject(err)
		end

		return resolve(connection)
	end)
end

return MessagingServiceUtils
